In [1]:
import sqlite3
import pandas as pd

# Setting up the database. DO NOT edit the code given below
from techsmart_db_setup import setup_database
setup_database() 
conn = sqlite3.connect('techsmart.db')

# Load and display tables
tables = ['Employee_Records', 'Product_Details', 'Customer_Demographics', 'Sales_Transactions']
for table in tables:
    query = f"SELECT * FROM {table} LIMIT 5"
    df = pd.read_sql_query(query, conn)
    print(f"\n{table}:")
    display(df)

✅ Database setup complete: Tables created and populated with data!

Employee_Records:


,employee_id,role,store_location,sales_performance
0,1,Cashier,New York,None
1,2,Cashier,Los Angeles,3000
2,3,Supervisor,Phoenix,None
3,4,Manager,Chicago,3000
4,5,Manager,Phoenix,3000



Product_Details:


,product_id,product_name,category,price,stock
0,106,Tablet,Electronics,one hundred,200.0
1,101,Keyboard,Accessories,fifty,100.0
2,108,Keyboard,Accessories,fifty,500.0
3,104,Speaker,Electronics,500,150.0
4,107,Speaker,Electronics,fifty,300.0



Customer_Demographics:


,customer_id,age,gender,location,loyalty_program
0,1,35,F,San Antonio,None
1,2,50,M,London,No
2,3,37,M,Austin,No
3,4,None,M,San Antonio,None
4,5,28,M,San Antonio,None



Sales_Transactions:


,transaction_id,customer_id,product_id,employee_id,quantity,total_amount,sale_date
0,1,13.0,105,2.0,4,90.0,2025-03-03 00:00:00
1,2,37.0,105,51.0,2,90.0,2025-03-08 00:00:00
2,3,5.0,106a,NaN,three,NaN,2025-03-01 00:00:00
3,4,35.0,109b,22.0,three,40.0,2025-03-04 00:00:00
4,5,15.0,101,45.0,2,NaN,2025-03-06 00:00:00


# TechMart Sales Analysis

## Project Overview

This project analyzes TechMart retail data to identify employee performance patterns, product sales trends, customer purchasing behavior, and store-category revenue performance.

The analysis uses SQL and Python to clean, transform, and analyze data across four related tables:

- Employee_Records
- Product_Details
- Customer_Demographics
- Sales_Transactions

## Business Objectives

The analysis focuses on four key business questions:

1. Analyze employee performance by location.
2. Identify top-selling products by category.
3. Analyze customer purchasing behavior.
4. Analyze Electronics and Accessories sales by employee, store, and customer loyalty status, while calculating total revenue by store-category pair for ranking and comparison.

## Data Preparation and Cleaning

Before conducting the analysis, the TechMart data was reviewed for missing values and inconsistent data formats across all four tables.

The cleaning process included:

- Identifying missing values.
- Identifying inconsistent data formats.
- Standardizing inconsistent values.
- Verifying the cleaned data before analysis.

## Dataset

The TechMart dataset contains four related tables representing employees, products, customers, and sales transactions.

| Table | Description |
|---|---|
| `Employee_Records` | Employee roles, store locations, and sales performance |
| `Product_Details` | Product names, categories, prices, and inventory levels |
| `Customer_Demographics` | Customer demographics and loyalty program status |
| `Sales_Transactions` | Sales transactions connecting customers, products, and employees |

Each table contains 100 records.

## Missing Values

The first stage of data preparation was to identify missing values across all four tables. This helps determine which fields require cleaning before conducting the business analysis.

### Missing Values: Employee Records

In [2]:
query = """
SELECT *
FROM Employee_Records
WHERE employee_id IS NULL
   OR role IS NULL
   OR store_location IS NULL
   OR sales_performance IS NULL;
"""

df = pd.read_sql_query(query, conn)
display(df)

,employee_id,role,store_location,sales_performance
0,1,Cashier,New York,None
1,3,Supervisor,Phoenix,None
2,26,Manager,New York,None
3,35,Sales Associate,Los Angeles,None
4,37,Cashier,London,None
5,45,Manager,Houston,None
6,46,Sales Associate,Los Angeles,None
7,68,Cashier,Los Angeles,None
8,69,Manager,London,None
9,76,Cashier,Los Angeles,None


In [3]:
query = """
SELECT
    SUM(CASE WHEN employee_id IS NULL THEN 1 ELSE 0 END) AS missing_employee_id,
    SUM(CASE WHEN role IS NULL THEN 1 ELSE 0 END) AS missing_role,
    SUM(CASE WHEN store_location IS NULL THEN 1 ELSE 0 END) AS missing_store_location,
    SUM(CASE WHEN sales_performance IS NULL THEN 1 ELSE 0 END) AS missing_sales_performance
FROM Employee_Records;
"""

df = pd.read_sql_query(query, conn)
display(df)

,missing_employee_id,missing_role,missing_store_location,missing_sales_performance
0,0,0,0,11


### Missing Values: Product Details

In [4]:
query = """
SELECT *
FROM Product_Details
WHERE product_id IS NULL
   OR product_name IS NULL
   OR category IS NULL
   OR price IS NULL
   OR stock IS NULL;
"""

df = pd.read_sql_query(query, conn)
display(df)

,product_id,product_name,category,price,stock
0,104,Tablet,Accessories,None,500.0
1,109,Mouse,Accessories,None,500.0
2,109,Keyboard,Electronics,fifty,NaN
3,104,Keyboard,Accessories,200,NaN
4,107,Mouse,Accessories,300,NaN
5,110,Speaker,Accessories,200,NaN
6,105,Mouse,Electronics,25,NaN
7,110,Monitor,Electronics,500,NaN
8,109,Keyboard,Electronics,fifty,NaN
9,110,Laptop,Electronics,one hundred,NaN


In [5]:
query = """
SELECT
    SUM(CASE WHEN product_id IS NULL THEN 1 ELSE 0 END) AS missing_product_id,
    SUM(CASE WHEN product_name IS NULL THEN 1 ELSE 0 END) AS missing_product_name,
    SUM(CASE WHEN category IS NULL THEN 1 ELSE 0 END) AS missing_category,
    SUM(CASE WHEN price IS NULL THEN 1 ELSE 0 END) AS missing_price,
    SUM(CASE WHEN stock IS NULL THEN 1 ELSE 0 END) AS missing_stock
FROM Product_Details;
"""

df = pd.read_sql_query(query, conn)
display(df)

,missing_product_id,missing_product_name,missing_category,missing_price,missing_stock
0,0,0,0,8,19


### Missing Values: Customer Demographics

In [6]:
query = """
SELECT *
FROM Customer_Demographics
WHERE customer_id IS NULL
   OR age IS NULL
   OR gender IS NULL
   OR location IS NULL
   OR loyalty_program IS NULL;
"""

df = pd.read_sql_query(query, conn)
display(df)

,customer_id,age,gender,location,loyalty_program
0,1,35,F,San Antonio,None
1,4,None,M,San Antonio,None
2,5,28,M,San Antonio,None
3,8,None,M,San Jose,No
4,10,41,F,Austin,None
5,13,37,F,London,None
6,14,None,F,San Antonio,No
7,16,22,M,New York,None
8,17,35,M,San Antonio,None
9,25,22,M,Los Angeles,None


In [7]:
query = """
SELECT
    SUM(CASE WHEN customer_id IS NULL THEN 1 ELSE 0 END) AS missing_customer_id,
    SUM(CASE WHEN age IS NULL THEN 1 ELSE 0 END) AS missing_age,
    SUM(CASE WHEN gender IS NULL THEN 1 ELSE 0 END) AS missing_gender,
    SUM(CASE WHEN location IS NULL THEN 1 ELSE 0 END) AS missing_location,
    SUM(CASE WHEN loyalty_program IS NULL THEN 1 ELSE 0 END) AS missing_loyalty_program
FROM Customer_Demographics;
"""

df = pd.read_sql_query(query, conn)
display(df)

,missing_customer_id,missing_age,missing_gender,missing_location,missing_loyalty_program
0,0,12,0,0,35


### Missing Values: Sales Transactions

In [8]:
query = """
SELECT *
FROM Sales_Transactions
WHERE transaction_id IS NULL
   OR customer_id IS NULL
   OR product_id IS NULL
   OR employee_id IS NULL
   OR quantity IS NULL
   OR total_amount IS NULL
   OR sale_date IS NULL;
"""

df = pd.read_sql_query(query, conn)
display(df)

,transaction_id,customer_id,product_id,employee_id,quantity,total_amount,sale_date
0,3,5.0,106a,NaN,three,NaN,2025-03-01 00:00:00
1,5,15.0,101,45.0,2,NaN,2025-03-06 00:00:00
2,8,13.0,102,39.0,three,NaN,2025-03-05 00:00:00
3,13,3.0,105,25.0,2,NaN,2025-03-04 00:00:00
4,15,NaN,109b,39.0,1,50.0,2025-03-05 00:00:00
5,17,NaN,103,NaN,2,NaN,2025-03-05 00:00:00
6,18,51.0,101,37.0,None,40.0,2025-03-04 00:00:00
7,19,3.0,101,37.0,three,NaN,2025-03-03 00:00:00
8,23,5.0,109b,37.0,None,50.0,2025-03-10 00:00:00
9,33,46.0,104,51.0,None,90.0,2025-03-10 00:00:00


In [9]:
query = """
SELECT
    SUM(CASE WHEN transaction_id IS NULL THEN 1 ELSE 0 END) AS missing_transaction_id,
    SUM(CASE WHEN customer_id IS NULL THEN 1 ELSE 0 END) AS missing_customer_id,
    SUM(CASE WHEN product_id IS NULL THEN 1 ELSE 0 END) AS missing_product_id,
    SUM(CASE WHEN employee_id IS NULL THEN 1 ELSE 0 END) AS missing_employee_id,
    SUM(CASE WHEN quantity IS NULL THEN 1 ELSE 0 END) AS missing_quantity,
    SUM(CASE WHEN total_amount IS NULL THEN 1 ELSE 0 END) AS missing_total_amount,
    SUM(CASE WHEN sale_date IS NULL THEN 1 ELSE 0 END) AS missing_sale_date
FROM Sales_Transactions;
"""

df = pd.read_sql_query(query, conn)
display(df)

,missing_transaction_id,missing_customer_id,missing_product_id,missing_employee_id,missing_quantity,missing_total_amount,missing_sale_date
0,0,7,0,7,17,16,0


## Inconsistent Data Formats

The datasets were also reviewed for values stored in inconsistent formats, particularly in columns expected to contain numeric values. These inconsistencies were standardized before performing the business analysis.

### Cleaning Employee Records

In [10]:
query = """
UPDATE Employee_Records
SET sales_performance = CASE
    WHEN sales_performance = 'five thousand' THEN '5000'
    WHEN sales_performance = 'nan' THEN NULL
    ELSE sales_performance
END;
"""

cursor = conn.cursor()
cursor.execute(query)
conn.commit()

In [11]:
query = """
SELECT *
FROM Employee_Records
WHERE sales_performance = 'five thousand'
   OR sales_performance = 'nan';
"""

df = pd.read_sql_query(query, conn)
display(df)

,employee_id,role,store_location,sales_performance


### Cleaning Product Details

In [12]:
query = """
UPDATE Product_Details
SET price = CASE
    WHEN price = 'one hundred' THEN '100'
    WHEN price = 'fifty' THEN '50'
    ELSE price
END;
"""

cursor.execute(query)
conn.commit()

In [13]:
query = """
SELECT *
FROM Product_Details
WHERE price = 'one hundred'
   OR price = 'fifty';
"""

df = pd.read_sql_query(query, conn)
display(df)

,product_id,product_name,category,price,stock


### Cleaning Customer Demographics

In [14]:
query = """
UPDATE Customer_Demographics
SET age = CASE
    WHEN age = 'thirty-five' THEN '35'
    ELSE age
END;
"""

cursor.execute(query)
conn.commit()

In [15]:
query = """
SELECT *
FROM Customer_Demographics
WHERE age = 'thirty-five';
"""

df = pd.read_sql_query(query, conn)
display(df)

,customer_id,age,gender,location,loyalty_program


### Cleaning Sales Transactions

In [16]:
query = """
UPDATE Sales_Transactions
SET
    quantity = CASE
        WHEN quantity = 'three' THEN '3'
        ELSE quantity
    END,
    product_id = CASE
        WHEN product_id = '106a' THEN '106'
        WHEN product_id = '109b' THEN '109'
        ELSE product_id
    END;
"""

cursor.execute(query)
conn.commit()

In [17]:
query = """
SELECT *
FROM Sales_Transactions
WHERE quantity = 'three'
   OR product_id IN ('106a', '109b');
"""

df = pd.read_sql_query(query, conn)
display(df)

,transaction_id,customer_id,product_id,employee_id,quantity,total_amount,sale_date


## Data Validation

After cleaning the inconsistent formats, the affected records were rechecked to confirm that the standardized values were applied successfully. The cleaned data can now be used for the business analysis.

In [18]:
query = """
SELECT *
FROM Employee_Records;
"""

df = pd.read_sql_query(query, conn)
display(df)

,employee_id,role,store_location,sales_performance
0,1,Cashier,New York,None
1,2,Cashier,Los Angeles,3000
2,3,Supervisor,Phoenix,None
3,4,Manager,Chicago,3000
4,5,Manager,Phoenix,3000
...,...,...,...,...
95,96,Manager,Phoenix,3000
96,97,Manager,London,1000
97,98,Manager,New York,2000
98,99,Sales Associate,Paris,3000


In [19]:
for table in [
    'Product_Details',
    'Customer_Demographics',
    'Sales_Transactions'
]:
    print(f"\n{table}:")
    df = pd.read_sql_query(f"SELECT * FROM {table} LIMIT 10", conn)
    display(df)


Product_Details:


,product_id,product_name,category,price,stock
0,106,Tablet,Electronics,100,200.0
1,101,Keyboard,Accessories,50,100.0
2,108,Keyboard,Accessories,50,500.0
3,104,Speaker,Electronics,500,150.0
4,107,Speaker,Electronics,50,300.0
5,104,Tablet,Accessories,None,500.0
6,109,Mouse,Accessories,None,500.0
7,109,Keyboard,Electronics,50,NaN
8,102,Charger,Electronics,25,500.0
9,106,Keyboard,Accessories,50,200.0



Customer_Demographics:


,customer_id,age,gender,location,loyalty_program
0,1,35,F,San Antonio,None
1,2,50,M,London,No
2,3,37,M,Austin,No
3,4,None,M,San Antonio,None
4,5,28,M,San Antonio,None
5,6,25,M,Los Angeles,Yes
6,7,43,F,New York,Yes
7,8,None,M,San Jose,No
8,9,42,M,Paris,No
9,10,41,F,Austin,None



Sales_Transactions:


,transaction_id,customer_id,product_id,employee_id,quantity,total_amount,sale_date
0,1,13.0,105,2.0,4,90.0,2025-03-03 00:00:00
1,2,37.0,105,51.0,2,90.0,2025-03-08 00:00:00
2,3,5.0,106,NaN,3,NaN,2025-03-01 00:00:00
3,4,35.0,109,22.0,3,40.0,2025-03-04 00:00:00
4,5,15.0,101,45.0,2,NaN,2025-03-06 00:00:00
5,6,4.0,103,25.0,1,40.0,2025-03-09 00:00:00
6,7,37.0,106,45.0,2,60.0,2025-03-06 00:00:00
7,8,13.0,102,39.0,3,NaN,2025-03-05 00:00:00
8,9,13.0,104,22.0,2,40.0,2025-03-04 00:00:00
9,10,42.0,101,15.0,2,60.0,2025-03-08 00:00:00


## SQL Skills Demonstrated

| SQL Skill | Examples Used |
|---|---|
| Filtering | `WHERE`, `IN`, `IS NOT NULL` |
| Aggregation | `SUM()`, `AVG()`, `COUNT()` |
| Grouping | `GROUP BY`, `HAVING` |
| Sorting | `ORDER BY` |
| Joins | `JOIN`, `LEFT JOIN` across multiple tables |
| Common Table Expressions | `WITH` |
| Window Functions | `RANK()`, `PARTITION BY` |
| Data Cleaning | `UPDATE`, `CASE` |
| Data Validation | Duplicate and unmatched-record checks |

## SQL Analysis

The cleaned TechMart data was analyzed to answer the key business questions defined for the project.

### 1. Employee Performance by Location

This analysis compares employee sales performance across store locations to identify locations with stronger overall employee performance.

In [20]:
query = """
SELECT
    store_location,
    COUNT(employee_id) AS employee_count,
    SUM(sales_performance) AS total_sales_performance,
    AVG(sales_performance) AS average_sales_performance
FROM Employee_Records
GROUP BY store_location
ORDER BY total_sales_performance DESC;
"""

df = pd.read_sql_query(query, conn)
display(df)

,store_location,employee_count,total_sales_performance,average_sales_performance
0,Phoenix,17,53000,3312.500000
1,New York,19,48000,2823.529412
2,London,16,45000,3214.285714
3,Los Angeles,18,35000,2692.307692
4,Chicago,8,31000,3875.000000
5,Paris,10,26000,2600.000000
6,Houston,12,26000,2363.636364


**Key Insight:**

Phoenix has the highest total employee sales performance at 53,000, followed by New York at 48,000 and London at 45,000.

However, Chicago has the highest average sales performance per employee at 3,875, despite having only 8 employees. This suggests that Chicago has fewer employees but stronger average individual performance.

### Employee Performance Ranking

Employees were ranked within each store location based on their sales performance.

In [21]:
query = """
SELECT
    employee_id,
    role,
    store_location,
    sales_performance,
    RANK() OVER (
        PARTITION BY store_location
        ORDER BY sales_performance DESC
    ) AS performance_rank
FROM Employee_Records
WHERE sales_performance IS NOT NULL
ORDER BY store_location, performance_rank;
"""

df = pd.read_sql_query(query, conn)
display(df)

,employee_id,role,store_location,sales_performance,performance_rank
0,17,Sales Associate,Chicago,5000,1
1,27,Supervisor,Chicago,5000,1
2,36,Manager,Chicago,5000,1
3,58,Sales Associate,Chicago,5000,1
4,78,Sales Associate,Chicago,5000,1
...,...,...,...,...,...
84,90,Manager,Phoenix,3000,5
85,91,Manager,Phoenix,3000,5
86,96,Manager,Phoenix,3000,5
87,43,Cashier,Phoenix,2000,15


**Employee Performance Insight:**

Employee performance varies within each store location. The ranking analysis identifies the highest-performing employees within each location while accounting for ties in sales performance.

### 2. Top-Selling Products by Category

This analysis identifies the top-selling products within each product category based on total quantity sold.

In [22]:
query = """
WITH product_sales AS (
    SELECT
        p.category,
        p.product_id,
        p.product_name,
        SUM(s.quantity) AS total_quantity_sold
    FROM Product_Details AS p
    JOIN Sales_Transactions AS s
        ON p.product_id = s.product_id
    WHERE s.quantity IS NOT NULL
    GROUP BY
        p.category,
        p.product_id,
        p.product_name
),
ranked_products AS (
    SELECT
        category,
        product_id,
        product_name,
        total_quantity_sold,
        RANK() OVER (
            PARTITION BY category
            ORDER BY total_quantity_sold DESC
        ) AS category_rank
    FROM product_sales
)
SELECT
    category,
    product_id,
    product_name,
    total_quantity_sold,
    category_rank
FROM ranked_products
WHERE category_rank = 1
ORDER BY category;
"""

df = pd.read_sql_query(query, conn)
display(df)

,category,product_id,product_name,total_quantity_sold,category_rank
0,Accessories,103,Keyboard,111,1
1,Electronics,103,Mouse,74,1
2,Electronics,103,Tablet,74,1


**Key Insight:**

The Keyboard is the top-selling product in the Accessories category, with 111 units sold.

In the Electronics category, the Mouse and Tablet are tied as the top-selling products, with 74 units sold each. This indicates that both products have similar sales demand and should be considered when planning Electronics inventory.

### 3. Customer Purchasing Behavior

This analysis examines customer purchasing frequency and total spending to identify high-value and frequent customers.

In [23]:
query = """
SELECT
    c.customer_id,
    c.location,
    c.loyalty_program,
    COUNT(s.transaction_id) AS total_transactions,
    SUM(s.total_amount) AS total_spent
FROM Customer_Demographics AS c
JOIN Sales_Transactions AS s
    ON c.customer_id = s.customer_id
WHERE s.total_amount IS NOT NULL
GROUP BY
    c.customer_id,
    c.location,
    c.loyalty_program
ORDER BY total_spent DESC;
"""

df = pd.read_sql_query(query, conn)
display(df)

,customer_id,location,loyalty_program,total_transactions,total_spent
0,13,London,None,8,440.0
1,12,San Antonio,Yes,6,400.0
2,55,San Jose,None,6,350.0
3,51,San Antonio,None,5,300.0
4,4,San Antonio,None,6,290.0
5,38,San Antonio,No,6,290.0
6,46,Phoenix,No,6,290.0
7,37,Chicago,None,6,280.0
8,5,San Antonio,None,5,270.0
9,3,Austin,No,4,220.0


**Key Insight:**

Customer 13 is the highest-spending customer in the dataset, with total spending of 440 across 8 transactions. Customer 12 follows with 400 across 6 transactions and is enrolled in the loyalty program.

Several high-spending customers are located in San Antonio, suggesting that this location may warrant further investigation. However, customers with missing loyalty-program information should not be classified as either members or non-members.

### 4. Electronics and Accessories Sales Analysis

This analysis focuses on Electronics and Accessories transactions by employee, store location, and customer loyalty status. The analysis compares transaction volume and revenue to identify sales patterns across these dimensions.

In [24]:
query = """
SELECT
    e.store_location,
    s.employee_id,
    c.loyalty_program,
    p.category,
    COUNT(s.transaction_id) AS total_transactions,
    SUM(s.quantity) AS total_quantity_sold,
    SUM(s.total_amount) AS total_revenue
FROM Sales_Transactions AS s
JOIN Product_Details AS p
    ON s.product_id = p.product_id
JOIN Customer_Demographics AS c
    ON s.customer_id = c.customer_id
JOIN Employee_Records AS e
    ON s.employee_id = e.employee_id
WHERE p.category IN ('Electronics', 'Accessories')
  AND s.quantity IS NOT NULL
  AND s.total_amount IS NOT NULL
GROUP BY
    e.store_location,
    s.employee_id,
    c.loyalty_program,
    p.category
ORDER BY total_revenue DESC;
"""

df = pd.read_sql_query(query, conn)
display(df)

,store_location,employee_id,loyalty_program,category,total_transactions,total_quantity_sold,total_revenue
0,Los Angeles,2.0,None,Electronics,26,89,1390.0
1,Los Angeles,2.0,None,Accessories,24,80,1320.0
2,Houston,45.0,No,Electronics,14,28,1060.0
3,Los Angeles,22.0,Yes,Electronics,17,37,1000.0
4,Phoenix,3.0,None,Electronics,16,60,940.0
5,New York,39.0,None,Electronics,21,55,920.0
6,Phoenix,3.0,None,Accessories,16,60,920.0
7,Houston,45.0,No,Accessories,12,24,880.0
8,Houston,45.0,None,Electronics,12,28,870.0
9,Los Angeles,15.0,Yes,Electronics,13,40,850.0


**Key Insight:**

Houston employee 45 generated the highest revenue among the displayed Electronics and Accessories combinations, generating 1,060 in Electronics revenue from 14 transactions and 880 in Accessories revenue from 12 transactions.

Los Angeles also shows strong Electronics performance, with employee 2 generating 970 in revenue from 19 transactions and employee 15 generating 850 from loyalty-program customers.

The results demonstrate differences in revenue across employees, store locations, product categories, and customer loyalty status. Missing loyalty-program values should be treated separately and not assumed to represent non-members.

### 5. Store-Category Revenue Ranking

This analysis calculates total revenue for each store and product-category combination and ranks the combinations from highest to lowest revenue.

In [25]:
query = """
WITH store_category_sales AS (
    SELECT
        e.store_location,
        p.category,
        SUM(s.total_amount) AS total_revenue
    FROM Sales_Transactions AS s
    JOIN Product_Details AS p
        ON s.product_id = p.product_id
    JOIN Employee_Records AS e
        ON s.employee_id = e.employee_id
    WHERE p.category IN ('Electronics', 'Accessories')
      AND s.total_amount IS NOT NULL
    GROUP BY
        e.store_location,
        p.category
)
SELECT
    store_location,
    category,
    total_revenue,
    RANK() OVER (
        ORDER BY total_revenue DESC
    ) AS revenue_rank
FROM store_category_sales
ORDER BY revenue_rank;
"""

df = pd.read_sql_query(query, conn)
display(df)

,store_location,category,total_revenue,revenue_rank
0,Los Angeles,Electronics,8120.0,1
1,Los Angeles,Accessories,7040.0,2
2,Houston,Electronics,2860.0,3
3,New York,Electronics,2740.0,4
4,Phoenix,Electronics,2430.0,5
5,Houston,Accessories,2410.0,6
6,New York,Accessories,2310.0,7
7,Phoenix,Accessories,2230.0,8
8,Chicago,Electronics,2130.0,9
9,London,Electronics,1770.0,10


**Key Insight:**

Los Angeles generated the highest revenue across both analyzed categories, with Electronics generating 5,510 and Accessories generating 4,190. Electronics was the highest-revenue category in every store, while London had the lowest revenue in both Electronics and Accessories.

The results indicate that Los Angeles is the strongest store location for Electronics and Accessories sales, while London may require further investigation to understand its comparatively lower revenue performance.

## Key Insights

The SQL analysis produced several key findings:

### Employee Performance
- Phoenix generated the highest total employee sales performance at 53,000.
- Chicago had the highest average employee performance at 3,875 despite having only 8 employees.

### Product Sales
- Keyboard was the top-selling Accessories product with 111 units sold.
- Mouse and Tablet tied as the top-selling Electronics products with 74 units each.

### Customer Purchasing Behavior
- Customer 13 was the highest-spending customer, with 440 in total spending across 8 transactions.
- Customer 12 was the second-highest spender at 400 and was enrolled in the loyalty program.
- Several high-spending customers were located in San Antonio.

### Electronics and Accessories
- Employee 45 in Houston generated 1,060 in Electronics revenue and 880 in Accessories revenue among the displayed results.
- Revenue varied across employees, locations, categories, and loyalty-program status.

### Store-Category Revenue
- Los Angeles Electronics generated the highest store-category revenue at 5,510.
- Los Angeles Accessories ranked second at 4,190.
- London Accessories had the lowest revenue at 960.

## Data Validation

Before finalizing the analysis, the cleaned tables were validated to check for duplicate identifiers and unmatched relationships between related tables.

In [26]:
query = """
SELECT
    employee_id,
    COUNT(*) AS record_count
FROM Employee_Records
GROUP BY employee_id
HAVING COUNT(*) > 1;
"""

df = pd.read_sql_query(query, conn)
display(df)

,employee_id,record_count


In [27]:
query = """
SELECT
    product_id,
    COUNT(*) AS record_count
FROM Product_Details
GROUP BY product_id
HAVING COUNT(*) > 1;
"""

df = pd.read_sql_query(query, conn)
display(df)

,product_id,record_count
0,101,7
1,102,10
2,103,10
3,104,9
4,105,8
5,106,6
6,107,15
7,108,5
8,109,15
9,110,15


In [28]:
query = """
SELECT
    customer_id,
    COUNT(*) AS record_count
FROM Customer_Demographics
GROUP BY customer_id
HAVING COUNT(*) > 1;
"""

df = pd.read_sql_query(query, conn)
display(df)

,customer_id,record_count


In [29]:
query = """
UPDATE Sales_Transactions
SET product_id = CASE
    WHEN product_id = '106a' THEN '106'
    WHEN product_id = '109b' THEN '109'
    ELSE product_id
END;
"""

cursor = conn.cursor()
cursor.execute(query)
conn.commit()

In [30]:
query = """
SELECT DISTINCT
    s.product_id
FROM Sales_Transactions AS s
LEFT JOIN Product_Details AS p
    ON s.product_id = p.product_id
WHERE p.product_id IS NULL;
"""

df = pd.read_sql_query(query, conn)
display(df)

,product_id


In [31]:
query = """
SELECT DISTINCT
    s.customer_id
FROM Sales_Transactions AS s
LEFT JOIN Customer_Demographics AS c
    ON s.customer_id = c.customer_id
WHERE s.customer_id IS NOT NULL
  AND c.customer_id IS NULL;
"""

df = pd.read_sql_query(query, conn)
display(df)

,customer_id


In [32]:
query = """
SELECT DISTINCT
    s.employee_id
FROM Sales_Transactions AS s
LEFT JOIN Employee_Records AS e
    ON s.employee_id = e.employee_id
WHERE s.employee_id IS NOT NULL
  AND e.employee_id IS NULL;
"""

df = pd.read_sql_query(query, conn)
display(df)

,employee_id


In [33]:
query = """
SELECT
    'Employee_Records' AS table_name,
    COUNT(*) AS remaining_inconsistent_values
FROM Employee_Records
WHERE sales_performance = 'five thousand'
   OR sales_performance = 'nan'

UNION ALL

SELECT
    'Product_Details',
    COUNT(*)
FROM Product_Details
WHERE price IN ('one hundred', 'fifty')

UNION ALL

SELECT
    'Customer_Demographics',
    COUNT(*)
FROM Customer_Demographics
WHERE age = 'thirty-five'

UNION ALL

SELECT
    'Sales_Transactions',
    COUNT(*)
FROM Sales_Transactions
WHERE quantity = 'three'
   OR product_id IN ('106a', '109b');
"""

df = pd.read_sql_query(query, conn)
display(df)

,table_name,remaining_inconsistent_values
0,Employee_Records,0
1,Product_Details,0
2,Customer_Demographics,0
3,Sales_Transactions,0


## Business Recommendations

Based on the analysis, the following recommendations can support TechMart's sales and operational decisions:

1. **Review high-performing locations:** Los Angeles demonstrates strong Electronics and Accessories revenue and could provide a benchmark for other stores.

2. **Optimize inventory by category:** Keyboard, Mouse, and Tablet show strong sales demand and should be monitored when planning inventory.

3. **Recognize high-performing employees:** Employee performance varies considerably across locations, so high-performing employees can provide useful benchmarks for training and performance management.

4. **Strengthen customer engagement:** High-spending and frequent customers can be targeted with loyalty and retention initiatives.

5. **Investigate lower-performing store-category combinations:** London, particularly Accessories, has comparatively low revenue and may require further analysis of demand, inventory, and customer purchasing patterns.

6. **Improve customer data quality:** Missing loyalty-program information should be reduced so that customer segmentation and loyalty analysis become more reliable.

## Conclusion

This project used SQL and Python to clean and analyze TechMart's employee, product, customer, and sales transaction data.

The analysis applied SQL techniques including filtering, aggregation, JOINs, CTEs, GROUP BY, subqueries, and window functions to answer business questions related to employee performance, product sales, customer purchasing behavior, and store-category revenue.

The findings provide actionable insights that can support inventory planning, employee performance management, customer engagement, and store-level sales strategy.